---
**Student Name**:

---



**Student ID:**

---
**Assignment 1:** Dynamic Programming

------


## **Question:** What is the purpose of OpenAI gyms and how is it going to help us in our RL education?

----

## **Answer:**

The OpenAI Gym is a toolkit (open source Python library) designed to provide a standardized environment for developing and testing reinforcement learning (RL) algorithms.

- Provides a common interface for different RL environments, making it easier to compare and benchmark different algorithms.
- Simplifies the process of creating and interacting with RL environments.
- Includes a diverse set of environments ranging from simple text-based environments to complex simulations.

--
- Implement RL algorithms and immediately test them in a controlled environment.
- experimentation with different algorithms and hyperparameters.
- Includes implementations of standard RL algorithms.


------


In [36]:
import gym

In [42]:
gym.envs.register(
    id='FrozenLakeNotSlippery-v0',
    entry_point='gym.envs.toy_text:FrozenLakeEnv',
    kwargs={'map_name' : '4x4', 'is_slippery': False},
    max_episode_steps=100,
    reward_threshold=0.74
)

In [43]:
# Create the gridworld-like environment
env=gym.make('FrozenLakeNotSlippery-v0')
# Let's look at the model of the environment (i.e., P):
env.env.P
# Question: what is the data in this structure saying? Relate this to the course
# presentation of P

{0: {0: [(1.0, 0, 0.0, False)],
  1: [(1.0, 4, 0.0, False)],
  2: [(1.0, 1, 0.0, False)],
  3: [(1.0, 0, 0.0, False)]},
 1: {0: [(1.0, 0, 0.0, False)],
  1: [(1.0, 5, 0.0, True)],
  2: [(1.0, 2, 0.0, False)],
  3: [(1.0, 1, 0.0, False)]},
 2: {0: [(1.0, 1, 0.0, False)],
  1: [(1.0, 6, 0.0, False)],
  2: [(1.0, 3, 0.0, False)],
  3: [(1.0, 2, 0.0, False)]},
 3: {0: [(1.0, 2, 0.0, False)],
  1: [(1.0, 7, 0.0, True)],
  2: [(1.0, 3, 0.0, False)],
  3: [(1.0, 3, 0.0, False)]},
 4: {0: [(1.0, 4, 0.0, False)],
  1: [(1.0, 8, 0.0, False)],
  2: [(1.0, 5, 0.0, True)],
  3: [(1.0, 0, 0.0, False)]},
 5: {0: [(1.0, 5, 0, True)],
  1: [(1.0, 5, 0, True)],
  2: [(1.0, 5, 0, True)],
  3: [(1.0, 5, 0, True)]},
 6: {0: [(1.0, 5, 0.0, True)],
  1: [(1.0, 10, 0.0, False)],
  2: [(1.0, 7, 0.0, True)],
  3: [(1.0, 2, 0.0, False)]},
 7: {0: [(1.0, 7, 0, True)],
  1: [(1.0, 7, 0, True)],
  2: [(1.0, 7, 0, True)],
  3: [(1.0, 7, 0, True)]},
 8: {0: [(1.0, 8, 0.0, False)],
  1: [(1.0, 12, 0.0, True)],
  2: [(

##**Question:** what is the data in this structure saying? Relate this to the course presentation of P

----

## **Answer**:

The provided structure for P in the FrozenLakeNotSlippery-v0 environment represents the transition dynamics of the MDP. (*book definition: The function p defines the dynamics of the MDP.*)

- **State** (0 to 15): Represents the states in the environment.
- **Action** (0 to 3): Represents the possible actions (0: Left, 1: Down, 2: Right, 3: Up).
- **Values** (List of tuples): Each tuple in the list describes a possible outcome: Probability (1.0): Since the environment is deterministic, the transition probability is always 1.0.
- **Next State:** The resulting state after taking the action.
- **Reward**: The reward received after the transition (1 if next state is 15, otherwise 0).
- **Is Terminal** (True/False): Indicates whether the episode ends after this transition (True if next state is terminal).

--

**Data Structure**

**Dictionary Format:**

*{(state, action): [(probability, next_state, reward, is_terminal)]}*

This shows the possible transitions for each state-action pair.

--

**Examples** (for explanation)

*State 0:*

- **Action 0 (Left):** [(1.0, 0, 0.0, False)] - Stays in state 0.
- **Action 1 (Down):** [(1.0, 4, 0.0, False)] - Moves to state 4.
- **Action 2 (Right):** [(1.0, 1, 0.0, False)] - Moves to state 1.
- **Action 3 (Up):** [(1.0, 0, 0.0, False)] - Stays in state 0.

*State 14:*

- **Action 2 (Right):** [(1.0, 15, 1.0, True)] - Moves to state 15 (the goal) with a reward of 1, ending the episode.

-----

In [4]:
# Now let's investigate the observation space (i.e., S using our nomenclature),
# and confirm we see it is a discrete space with 16 locations
print(env.observation_space)

Discrete(16)


In [5]:
stateSpaceSize = env.observation_space.n
print(stateSpaceSize)

16


In [6]:
# Now let's investigate the action space (i.e., A) for the agent->environment
# channel
print(env.action_space)

Discrete(4)


In [8]:
# The gym environment has ...sample() functions that allow us to sample
# from the above spaces:
for g in range(1,10,1):
  print("sample from S:",env.observation_space.sample()," ... ","sample from A:",env.action_space.sample())

sample from S: 7  ...  sample from A: 1
sample from S: 15  ...  sample from A: 2
sample from S: 13  ...  sample from A: 0
sample from S: 0  ...  sample from A: 0
sample from S: 11  ...  sample from A: 1
sample from S: 10  ...  sample from A: 3
sample from S: 15  ...  sample from A: 3
sample from S: 1  ...  sample from A: 0
sample from S: 9  ...  sample from A: 2


In [ ]:
# The enviroment also provides a helper to render (visualize) the environment
env.reset()
env.render()

In [12]:
# We can act as the agent, by selecting actions and stepping the environment
# through time to see its responses to our actions
env.reset()
exitCommand=False
while not(exitCommand):
  env.render()
  print("Enter the action as an integer from 0 to",env.action_space.n," (or exit): ")
  userInput=input()
  if userInput=="exit":
    break
  action=int(userInput)
  (observation, reward, compute, probability) = env.step(action)
  print("--> The result of taking action",action,"is:")
  print("     S=",observation)
  print("     R=",reward)
  print("     p=",probability)

  env.render()


Enter the action as an integer from 0 to 4  (or exit): 
3
--> The result of taking action 3 is:
     S= 0
     R= 0.0
     p= {'prob': 1.0}
Enter the action as an integer from 0 to 4  (or exit): 
2
--> The result of taking action 2 is:
     S= 1
     R= 0.0
     p= {'prob': 1.0}
Enter the action as an integer from 0 to 4  (or exit): 
1
--> The result of taking action 1 is:
     S= 5
     R= 0.0
     p= {'prob': 1.0, 'TimeLimit.truncated': False}
Enter the action as an integer from 0 to 4  (or exit): 
exit


-----
## **Question:** draw a table indicating the correspondence between the action you input (a number) and the logic action performed.

In [20]:
'''
Question: draw a table indicating the correspondence between the action
you input (a number) and the logic action performed.
------------------------------------------------------------------------
'''

import warnings

# to hide warnings
warnings.filterwarnings('ignore', category=DeprecationWarning)

#for better tabular visualzation
import pandas as pd

actions= {
    "Action input": [0, 1, 2, 3],
    "Logic action perfomed": ["Left", "Down", "Right", "Up"]
}

actions_table = pd.DataFrame(actions)
actions_table

Action input Logic action perfomed
0             0                  Left
1             1                  Down
2             2                 Right
3             3                    Up

---
## **Question**: draw a table that illustrates what the symbols on the render image mean?

In [22]:

symbols_data = {'Symbols':['S', 'H', 'F', 'G'],
        'Meaning':['Start', 'Hole', 'Frozen', 'Goal']}

df = pd.DataFrame(symbols_data)
df


Symbols Meaning
0       S   Start
1       H    Hole
2       F  Frozen
3       G    Goal

---
## **Question:** Explain what the objective of the agent is in this environment?


## **Answer:**



In the FrozenLake environment, the agent's objective is to navigate from the starting point (S) to the goal point (G) while avoiding falling into the holes (H). The agent receives a reward of 1 upon successfully reaching the goal and a reward of 0 if it falls into a hole.

- The agent must learn to navigate the environment in such a way that it avoids falling into any of the holes. Falling into a hole results in an immediate loss and zero reward.

- The agent's ultimate objective is to reach the goal point while maximizing its reward. Upon reaching the goal, the agent receives a reward of 1, indicating successful completion of the task.

- Along the path from the starting point to the goal, the agent should aim to accumulate as much reward as possible by making efficient and safe moves.

*https://www.gymlibrary.dev/environments/toy_text/frozen_lake/*

----------
## **Practical:** Code up an AI that will employ random action selection in order to drive the agent. Test this random action selection agent with the above environment (i.e., code up a loop as I did above, but instead of taking input from a human user, take it from the AI you coded).


In [39]:
env.reset()
# Run the episode until termination
while True:
  env.render()

  # Select a random action from the action space
  # (https://blog.paperspace.com/getting-started-with-openai-gym/)

  action = env.action_space.sample()

  observation, reward, done, info = env.step(action)

  # Print the result of taking the action
  print("--> The result of taking action", action, "is:")
  print("     S=", observation)
  print("     R=", reward)
  print("     Done=", done)

  # If the episode is done, break out of the loop
  if done:
        # If the agent fell into a hole
    if reward == 0:
        print("Agent fell into a hole.")
    # If the agent reached the target
    elif reward == 1:
        print("Agent reached the target.")
    break

  env.render()

--> The result of taking action 3 is:
     S= 0
     R= 0.0
     Done= False
--> The result of taking action 2 is:
     S= 1
     R= 0.0
     Done= False
--> The result of taking action 2 is:
     S= 2
     R= 0.0
     Done= False
--> The result of taking action 3 is:
     S= 2
     R= 0.0
     Done= False
--> The result of taking action 3 is:
     S= 2
     R= 0.0
     Done= False
--> The result of taking action 3 is:
     S= 2
     R= 0.0
     Done= False
--> The result of taking action 0 is:
     S= 1
     R= 0.0
     Done= False
--> The result of taking action 3 is:
     S= 1
     R= 0.0
     Done= False
--> The result of taking action 1 is:
     S= 5
     R= 0.0
     Done= True
Agent fell into a hole.


In [ ]:
# Now towards dynamic programming. Note that env.env.P has the model
# of the environment.
#
# Question: How would you represent the agent's policy function and value function?
# Practical: revise the above AI solver to use a policy function in which you
# code the random action selections in the policy function. Test this.
# Practical: Code the C-4 Policy Evaluation (Prediction) algorithm. You may use
# either the inplace or ping-pong buffer (as described in the lecture). Now
# randomly initialize your policy function, and compute its value function.
# Report your results: policy and value function. Ensure your prediction
# algo reports how many iterations it took.
#
# (Optional): Repeat the above for q.
#
# Policy Improvement:
# Question: How would you use P and your value function to improve an arbitrary
# policy, pi, per Chapter 4?
# Practical: Code the policy iteration process, and employ it to arrive at a
# policy that solves this problem. Show your testing results, and ensure
# it reports the number of iterations for each step: (a) overall policy
# iteration steps and (b) evaluation steps.
# Practical: Code the value iteration process, and employ it to arrive at a
# policy that solves this problem. Show your testing results, reporting
# the iteration counts.
# Comment on the difference between the iterations required for policy vs
# value iteration.
#
# Optional: instead of the above environment, use the "slippery" Frozen Lake via
# env = gym.make("FrozenLake-v0")

------------------
## **Question:** How would you represent the agent's policy function and value function?

## **Answer:**

- We start with a set of probabilities $p(s',r|s,a)$, which represent the dynamics of the environment, indicating the probability of transitioning to state $s'$ and receiving reward $r$ given action $a$ in state $s$.

- Using the Bellman equation for the state value function $v_{𝜋}$, we iteratively update the value of each state based on the current policy $𝜋$:

  ($v_{k+1}(s)=\sum_{a}\pi(a|s)\sum_{s',r}p(s',r|s,a)[r+\gamma v_{k}(s')]$).

- At each iteration, we estimate the expected return from each state by considering the possible actions, transitions to next states, and rewards, weighted by the policy probabilities and the transition dynamics.

- The resulting approximate value function $v_{k+1}$ provides an estimate of the expected return from each state under the current policy  $𝜋$

----
## **Practical:** revise the above AI solver to use a policy function in which you code the random action selections in the policy function. Test this.



In [44]:
import numpy as np
import random

def random_action_policy(stateSpaceSize, actionSpaceSize):
    return np.ones([stateSpaceSize, actionSpaceSize]) / actionSpaceSize


env.reset()
while True:
    env.render()

    # Get the current observation (state) from the environment
    observation = env.env.s

    # Generate a random action policy for the current state space and action space
    policy = random_action_policy(env.observation_space.n, env.action_space.n)

    # Choose action according to the generated policy
    action = np.random.choice(env.action_space.n, p=policy[observation])

    # Take the chosen action in the environment
    (observation, reward, done, probability) = env.step(action)

    # Print the result of taking the action
    print("--> The result of taking action", action, "is:")
    print("     S=", observation)
    print("     R=", reward)
    print("     p=", probability)

    if done and reward == 1:
        print("Goal Reached")
        break
    if done and reward == 0:
        print("Agent fell into Hole")
        break

    # Render the updated state of the environment
    env.render()



--> The result of taking action 1 is:
     S= 4
     R= 0.0
     p= {'prob': 1.0}
--> The result of taking action 3 is:
     S= 0
     R= 0.0
     p= {'prob': 1.0}
--> The result of taking action 0 is:
     S= 0
     R= 0.0
     p= {'prob': 1.0}
--> The result of taking action 0 is:
     S= 0
     R= 0.0
     p= {'prob': 1.0}
--> The result of taking action 1 is:
     S= 4
     R= 0.0
     p= {'prob': 1.0}
--> The result of taking action 1 is:
     S= 8
     R= 0.0
     p= {'prob': 1.0}
--> The result of taking action 0 is:
     S= 8
     R= 0.0
     p= {'prob': 1.0}
--> The result of taking action 1 is:
     S= 12
     R= 0.0
     p= {'prob': 1.0, 'TimeLimit.truncated': False}
Agent fell into Hole


---
## **Practical**: Code the C-4 Policy Evaluation (Prediction) algorithm. You may use either the inplace or ping-pong buffer (as described in the lecture). Now randomly initialize your policy function, and compute its value function. Report your results: policy and value function. Ensure your prediction algo reports how many iterations it took.

In [50]:
def random_policy(env):
    return np.random.randint(env.action_space.n, size=env.observation_space.n)

def policy_evaluation(env, policy, gamma=0.5, theta=1e-6):

    V_current = np.zeros(env.observation_space.n)
    V_new = np.zeros(env.observation_space.n)

    num_iterations = 0

    while True:
        delta = 0

        for s in range(env.observation_space.n):
            v = V_current[s]
            action = policy[s]
            transition_probs = env.P[s][action]

            # Calculate the new value of the state using the Bellman
            v_new = sum(prob * (reward + gamma * V_current[next_state]) for prob, next_state, reward, _ in transition_probs)

            # Update the maximum change in the value function
            delta = max(delta, abs(v - v_new))

            # the new buffer
            V_new[s] = v_new

        # Swap the buffers
        V_current, V_new = V_new, V_current

        num_iterations += 1  # Increment iteration counter

        # convergence Check
        if delta < theta:
            break

    return V_current, num_iterations



env.reset()

policy = random_policy(env)

value_function, num_iterations = policy_evaluation(env, policy)

print("\nValue Function:")
print(value_function)
print("\nNumber of Iterations:", num_iterations)



Value Function:
[1.62904666e-04 0.00000000e+00 0.00000000e+00 0.00000000e+00
 8.15210584e-04 0.00000000e+00 0.00000000e+00 0.00000000e+00
 4.07666300e-03 2.36452599e-02 6.89654870e-02 0.00000000e+00
 0.00000000e+00 7.29063367e-02 4.13793055e-01 0.00000000e+00]

Number of Iterations: 12


-----
## **(Optional):** Repeat the above for q.

In [54]:
def q_value_iteration(env, gamma=1.0, theta=1e-6):
    Q = np.zeros((env.observation_space.n, env.action_space.n))

    num_iterations = 0
    while True:
        delta = 0
        for s in range(env.observation_space.n):
            for a in range(env.action_space.n):
                q = Q[s][a]

                # Update Q-values using Bellman
                next_states = env.P[s][a]

                new_q = sum(prob * (reward + gamma * max(Q[next_state])) for prob, next_state, reward, _ in next_states)
                Q[s][a] = new_q

                delta = max(delta, abs(q - Q[s][a]))

        num_iterations += 1

        if delta < theta:
            break
    return Q, num_iterations

env.reset()

Q_values, num_iterations = q_value_iteration(env)

# Results
print("Q-values:")
print(Q_values)
print("\nNumber of Iterations:", num_iterations)

Q-values:
[[0.82351228 0.82351054 0.82351054 0.82351014]
 [0.54900605 0.5490047  0.54900256 0.82350666]
 [0.72546939 0.7254687  0.72546709 0.82350275]
 [0.54900084 0.54900084 0.54899985 0.82350076]
 [0.823514   0.54901009 0.54900951 0.54900876]
 [0.         0.         0.         0.        ]
 [0.52940008 0.25489917 0.52940008 0.27450092]
 [0.         0.         0.         0.        ]
 [0.54901009 0.54901203 0.54901128 0.8235167 ]
 [0.568621   0.82352016 0.5490146  0.52940473]
 [0.76469778 0.58823108 0.49019106 0.45097341]
 [0.         0.         0.         0.        ]
 [0.         0.         0.         0.        ]
 [0.56862215 0.60783979 0.88234651 0.58823108]
 [0.86273913 0.9411732  0.90195699 0.8823481 ]
 [0.         0.         0.         0.        ]]

Number of Iterations: 319


---

## **Policy Improvement:**

**Question:** How would you use P and your value function to improve an arbitrary policy, pi, per Chapter 4?

**Answer:**

- Compute the action-value function $q_{\pi}(s, a)$ for all actions a using the Bellman equation: $q_{\pi}(s, a) = \sum_{s', r} p(s', r | s, a) [r + \gamma v_{\pi}(s')]$

- Update the policy $\pi'(s)$ by selecting the action that maximizes the action-value function: $\pi'(s) = \arg\max_{a} q_{\pi}(s, a)$

- Compute the value function $v_{\pi'}(s)$ for the current policy $\pi'$ using the Bellman expectation equation:
$v_{\pi'}(s) = \sum_{a} \pi'(a|s) \sum_{s', r} p(s', r|s, a) [r + \gamma v_{\pi'}(s')]$

- Improve the policy using the value function to obtain a new policy
- Update the value function $v$ using the Bellman optimality equation:

----
## **Practical**: Code the policy iteration process, and employ it to arrive at a policy that solves this problem. Show your testing results, and ensure it reports the number of iterations for each step: (a) overall policy iteration steps and (b) evaluation steps.

In [84]:
def policy_evaluation(env, policy, gamma=1.0, theta=1e-6, max_iterations=10000):
    num_states = env.observation_space.n
    value_function = np.zeros(num_states)
    num_iterations = 0

    for _ in range(max_iterations):
        delta = 0
        for state in range(num_states):
            v = value_function[state]
            action = policy[state]
            state_value = 0
            for prob, next_state, reward, _ in env.P[state][action]:
                state_value += prob * (reward + gamma * value_function[next_state])
            value_function[state] = state_value
            delta = max(delta, abs(v - value_function[state]))
        num_iterations += 1
        if delta < theta:
            break

    return value_function, num_iterations

def policy_improvement(env, value_function, gamma=1.0):

    num_states = env.observation_space.n
    num_actions = env.action_space.n
    policy = np.zeros(num_states, dtype=int)

    for state in range(num_states):
        action_values = np.zeros(num_actions)
        for action in range(num_actions):
            for prob, next_state, reward, _ in env.P[state][action]:
                action_values[action] += prob * (reward + gamma * value_function[next_state])
        best_action = np.argmax(action_values)
        policy[state] = best_action

    return policy

def policy_iteration(env, gamma=1.0, theta=1e-6, max_iterations=10000):

    num_states = env.observation_space.n
    num_actions = env.action_space.n

    # Initialize a random policy
    policy = np.random.randint(0, num_actions, num_states)
    num_policy_iterations = 0
    num_evaluation_iterations = 0

    while True:
        num_policy_iterations += 1

        # Policy Evaluation
        value_function, num_iterations = policy_evaluation(env, policy, gamma, theta, max_iterations)
        num_evaluation_iterations += num_iterations

        # Policy Improvement
        new_policy = policy_improvement(env, value_function, gamma)

        # Check for convergence
        if np.array_equal(new_policy, policy):
            break

        policy = new_policy

    return policy, num_policy_iterations, num_evaluation_iterations

env.reset()

# Perform policy iteration
optimal_policy, num_policy_iterations, num_evaluation_iterations = policy_iteration(env)

print("Optimal Policy:", optimal_policy)
print("Number of Policy Iterations:", num_policy_iterations)
print("Number of Evaluation Iterations:", num_evaluation_iterations)


Optimal Policy: [0 3 3 3 0 0 0 0 3 1 0 0 0 2 1 0]
Number of Policy Iterations: 4
Number of Evaluation Iterations: 864


---
## **Practical**: Code the value iteration process, and employ it to arrive at a policy that solves this problem. Show your testing results, reporting the iteration counts.

In [86]:
def value_iteration(env, gamma=1.0, theta=1e-6, max_iterations=10000):

    num_states = env.observation_space.n
    num_actions = env.action_space.n
    value_function = np.zeros(num_states)
    num_iterations = 0

    while True:
        delta = 0
        for state in range(num_states):
            v = value_function[state]
            action_values = np.zeros(num_actions)
            for action in range(num_actions):
                for prob, next_state, reward, _ in env.P[state][action]:
                    action_values[action] += prob * (reward + gamma * value_function[next_state])
            value_function[state] = np.max(action_values)
            delta = max(delta, np.abs(v - value_function[state]))
        num_iterations += 1
        if delta < theta or num_iterations >= max_iterations:
            break

    # Extract the optimal policy
    optimal_policy = np.zeros(num_states, dtype=int)
    for state in range(num_states):
        action_values = np.zeros(num_actions)
        for action in range(num_actions):
            for prob, next_state, reward, _ in env.P[state][action]:
                action_values[action] += prob * (reward + gamma * value_function[next_state])
        optimal_policy[state] = np.argmax(action_values)

    return optimal_policy, num_iterations

env.reset()

optimal_policy, num_iterations = value_iteration(env)

print("Optimal Policy:", optimal_policy)
print("Number of Iterations:", num_iterations)

Optimal Policy: [0 3 3 3 0 0 0 0 3 1 0 0 0 2 1 0]
Number of Iterations: 321


----
## Comment on the difference between the iterations required for policy vs value iteration.



**Answer:**

- Policy Iteration: Alternates between policy evaluation and improvement. Each policy iteration involves a complete evaluation of the current policy, which can lead to a higher number of iterations.
- Value Iteration: Directly computes the optimal value function by iteratively updating the value estimates for each state until convergence. It typically requires fewer iterations compared to policy iteration since it does not involve separate policy evaluation steps.
- In the results, policy iteration required 864 evaluations, while value iteration required 321 iterations to converge to the optimal policy.
